java-checks/src/main/java/org/sonar/java/checks/DateTimeFormatterMismatchCheck.java

```diff
@@ -0,0 +1,164 @@
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S5917")
+public class DateTimeFormatterMismatchCheck extends IssuableSubscriptionVisitor {
+
+  private static final MethodMatchers APPEND_VALUE_MATCHER = MethodMatchers.create()
+    .ofTypes("java.time.format.DateTimeFormatterBuilder")
+    .names("appendValue")
+    .addParametersMatcher("java.time.temporal.TemporalField", "int")
+    .build();
+
+  private static final MethodMatchers WEEK_BASED_YEAR_MATCHER = MethodMatchers.create()
+    .ofTypes("java.time.temporal.WeekFields")
+    .names("weekBasedYear")
+    .addWithoutParametersMatcher()
+    .build();
+
+  private static final MethodMatchers WEEK_OF_WEEK_BASED_YEAR_MATCHER = MethodMatchers.create()
+    .ofTypes("java.time.temporal.WeekFields")
+    .names("weekOfWeekBasedYear")
+    .addWithoutParametersMatcher()
+    .build();
+
+  private static final String CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE = "Change this year format to use the week-based year instead.";
+  private static final String CHANGE_YEAR_FORMAT_TO_CHRONOFIELD_MESSAGE = "Change this year format to use ChronoField.YEAR instead.";
+  private static final String SECONDARY_LOCATION_MESSAGE = "";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree invocation = (MethodInvocationTree) tree;
+    if (APPEND_VALUE_MATCHER.matches(invocation)) {
+      ChainedInvocationVisitor visitor = new ChainedInvocationVisitor();
+      invocation.accept(visitor);
+      if (!visitor.usesWeek || !visitor.usesYear) {
+        return;
+      }
+      if (visitor.usesWeekBasedYear && !visitor.usesWeekOfWeekBasedYear) {
+        reportIssue(visitor.primary, CHANGE_YEAR_FORMAT_TO_CHRONOFIELD_MESSAGE, visitor.secondaries, null);
+      } else if (!visitor.usesWeekBasedYear && visitor.usesWeekOfWeekBasedYear) {
+        reportIssue(visitor.primary, CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE, visitor.secondaries, null);
+      }
+    }
+  }
+
+  private static class ChainedInvocationVisitor extends BaseTreeVisitor {
+    private boolean usesWeek = false;
+    private boolean usesWeekOfWeekBasedYear = false;
+    private boolean usesYear = false;
+    private boolean usesWeekBasedYear = false;
+    private final List<JavaFileScannerContext.Location> secondaries = new ArrayList<>();
+    private ExpressionTree primary = null;
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree callToAppendValue) {
+      MethodInvocationTree tree = callToAppendValue;
+      inspectCall(tree);
```

Comment:
By going through the tree top-down instead of bottom-up, we can get rid of the loop and calls to `subsequentMethodInvocation` here and instead just recursively visit the child nodes.

Reply:
The original implementation used to start from `toFormatter` but that would prevent the check from detecting conflicts if the `appendValue` calls are not chained with it. Similarly, we could take a look at the builder's constructor but we would end up in the same situation.
To prevent this, we now identify the entire chain and only then process it looking for conflicts